### PR TITLE
Steppable Optimizers

### PR DESCRIPTION
**Pull request by MarcDrudis**
### Summary
We incorporate a new subclass of Optimizer called SteppableOptimizer. A steppable optimizer has a method step() that allows to perform one single step in the optimization process as opposed to having to perform the whole minimization of the objective function.
This way the user has more options when dealing with complicated functions to minimize.
In this spirit, we have chosen to add ask() and tell() to this interface as well. More information about this can be found in the documentation.

The main goal of this PR is to implement the interface of the abstract class SteppableOptimizer and document it properly. In order to make some tests, we have implemented GradientDescent using this new framework.

### Details and comments
A class CMA-ES is also in the making, but I prefered to start by making this PR with GradientDescent because it is easier to implement and later on we can add this other class.

### CODE REVIEW

File: qiskit/algorithms/optimizers/gradient_descent.py [hunk 1]
```diff
@@ -149,60 +60,104 @@ def settings(self) -> Dict[str, Any]:
...
+        When the stepsize is smaller than the tolerance, the optimization process is considered
+        finished.
+        Returns:
+            True if the optimization process should continue, False otherwise.
+        """
 
-        return result
+        cont_condition = self._state.stepsize > self.tol
+        cont_condition &= super().continue_condition()
+        return cont_condition
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/gradient_descent.py:160]: Just a suggestion you're welcome to ignore, but maybe this would be more concise? 
```suggestion
        return cont_condition and super().continue_condition()
```

File: qiskit/algorithms/optimizers/gradient_descent.py [hunk 2]
```diff
@@ -1,136 +1,47 @@
...
     def __init__(
         self,
-        maxiter: int = 100,
+        maxiter: int = 1000,
+        callback: Optional[CALLBACK] = None,
         learning_rate: Union[float, Callable[[], Iterator]] = 0.01,
         tol: float = 1e-7,
-        callback: Optional[CALLBACK] = None,
         perturbation: Optional[float] = None,
+        **kwargs,
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/gradient_descent.py:37]: It seems like this one is not required
```suggestion
```

File: qiskit/algorithms/optimizers/gradient_descent.py [hunk 3]
```diff
@@ -1,136 +1,47 @@
...
 
+class GradientDescent(SteppableOptimizer):
     def __init__(
         self,
-        maxiter: int = 100,
+        maxiter: int = 1000,
+        callback: Optional[CALLBACK] = None,
         learning_rate: Union[float, Callable[[], Iterator]] = 0.01,
         tol: float = 1e-7,
-        callback: Optional[CALLBACK] = None,
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/gradient_descent.py:112]: Could we keep the original default values and order of the arguments? 🙂

File: qiskit/algorithms/optimizers/gradient_descent.py [hunk 4]
```diff
@@ -1,136 +1,47 @@
...
 
-            initial_point = np.array([1, 0.5, -0.2])
 
-            optimizer = GradientDescent(maxiter=100, learning_rate=learning_rate)
-            x_opt, fx_opt, nfevs = optimizer.optimize(initial_point.size,
-                                                      f,
-                                                      gradient_function=grad_f,
-                                                      initial_point=initial_point)
+@dataclass
+class GD_OptimizerState(OptimizerState):
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/gradient_descent.py:23]: Could we rename this e.g. to `GradientDescentState`? This would fit our naming scheme of writing things out and not using underscores a bit nicer 🙂

File: qiskit/algorithms/optimizers/gradient_descent.py [hunk 5]
```diff
@@ -149,60 +60,104 @@ def settings(self) -> Dict[str, Any]:
...
+        This is the default way of evaluating the function given the request by self.ask().
+        For gradient descent we are going to check how to evaluate the gradient, evaluate and
+        return a TellObject.
+        """
+        if self._state.jac is None:
+            grad = Optimizer.gradient_num_diff(
+                x_center=ask_object.x_fun,
+                f=self._state.fun,
+                epsilon=self.perturbation,
+                max_evals_grouped=1,  # Here there was some extra logic I am just neglecting for now.
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/gradient_descent.py:101]: Is this comment still open?

File: qiskit/algorithms/optimizers/gradient_descent.py [hunk 6]
```diff
@@ -149,60 +60,104 @@ def settings(self) -> Dict[str, Any]:
...
+            stepsize=np.inf,
+        )
 
-        # TODO the optimizer result should contain the number of gradient evaluations,
-        # if the gradient is passed
-        result = OptimizerResult()
-        result.x = x
-        result.fun = fun(x)
-        result.nfev = nfevs
+    def continue_condition(self) -> bool:
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/gradient_descent.py:149]: I'm not 100% happy with this name yet... I feel like `terminate` or `termination_criterion` is more standard. If we really want to use "continue" could we just drop the `_condition`? That'd be a bit shorter and it's likely clear from context what this function means 🙂

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
+from abc import abstractmethod, ABC
+from dataclasses import dataclass, field
+from typing import Dict, Any, Union, Callable, Optional, Tuple, List
+
+from .optimizer import Optimizer, POINT, OptimizerResult
+import numpy as np
+
+CALLBACK = Callable[[int, np.ndarray, float, float], None]
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:8]: I believe the callback differs for the optimizers (see e.g. gradient descent and SPSA), probably it makes more sense to just keep the definitions inside the files of the optimizers.

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    """
+    Base class for return type of method :meth:`~qiskit.algorithms.optimizers.SteppableOptimizer.ask`.
+    Args:
+        x_fun: Point or list of points where the function needs to be evaluated to compute the
+            next state of the optimizer.
+        x_jac: Point or list of points where the gradient/jacobian needs to be evaluated to compute the
+            next state of the optimizer.
+    """
+
+    x_fun: Union[POINT, List[POINT], None] = None
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:22]: Typically we use `Optional[...]` instead of `Union[..., None]` (which is the exact purpose of `Optional`)
```suggestion
    x_fun: Optional[Union[POINT, List[POINT]]] = None
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        def grad(x):
+            if random.choice([True, False]):
+                return None
+            else:
+                return 2 * (np.linalg.norm(x) - 1) * x / np.linalg.norm(x)
+
+
+        initial_point = np.random.normal(0, 1, size=(100,))
+
+        optimizer = SteppableGradientDescent(maxiter=20)
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:93]:
```suggestion
        optimizer = GradientDescent(maxiter=20)
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    Once the function has been evaluated, the user calls the method tell() to tell the optimizer what has been the result of the function evaluation.
+    The optimizer then updates its state accordingly and the user can decide whether to stop the optimization process or to repeat a step.
+
+    This interface is more customizable, and allows the user to have full control over the evaluation of the function.
+
+    For example:
+    .. code-block::python
+        import random
+        import numpy as np
+        from qiskit.algorithms.optimizers import SteppableGradientDescent
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:76]:
```suggestion
        from qiskit.algorithms.optimizers import GradientDescent
```

File: qiskit/algorithms/optimizers/gradient_descent.py (hunk 3 above)
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/gradient_descent.py:112]: Yes, this is actually a very good point. I absolutely need to make sure that we don't make any old code obsolete by some kind of mistake of this kind.

File: qiskit/algorithms/optimizers/gradient_descent.py [hunk 11]
```diff
@@ -1,136 +1,47 @@
...
-            optimizer = GradientDescent(maxiter=100, learning_rate=learning_rate)
-            x_opt, fx_opt, nfevs = optimizer.optimize(initial_point.size,
-                                                      f,
-                                                      gradient_function=grad_f,
-                                                      initial_point=initial_point)
+@dataclass
+class GD_OptimizerState(OptimizerState):
 
-            print(f"Found minimum {x_opt} at a value of {fx_opt} using {nfevs} evaluations.")
+    eta: Union[float, Callable[[], Iterator]]
```
[the remark this answers is not in the document]
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/gradient_descent.py:25]: Is it a typo that you wrote learning_rate instead of eta?
This comes from the last implementation and might be a bit convoluted but I will try to explain it.
learning_rate can be a function that takes no arguments and returns an iterator. In that case eta will become the iterator returned by that function and can be used to generate all of the floats corresponding to the "learning rate"(the quantity not the variable) on each step of the optimization.
learning_rate can also be a float. In that case eta becomes an iterator that always yields that value.

learning_rate and eta are not exactly the same. Think of learning rate as a factory of iterators, and eta as an iterator. Since it can only be used once we need to restart it if we want to perform an other optimization.

However, I mistakenly left `Union[float,Callable[[],Iterator]` as typehint of eta when in reality eta is always an `Iterator`. I have changed it now. Thank you!

File: qiskit/algorithms/optimizers/gradient_descent.py (hunk 4 above)
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/gradient_descent.py:23]: Okay, thank you!

File: qiskit/algorithms/optimizers/gradient_descent.py (hunk 5 above)
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/gradient_descent.py:101]: Yes, this value is not 1 in the old gradient descent class, but I haven't had time to look what it makes reference to.

File: qiskit/algorithms/optimizers/gradient_descent.py (hunk 6 above)
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/gradient_descent.py:149]: We can discuss on next meeting whether we want to make it something positive or something negative.

File: qiskit/algorithms/optimizers/steppable_optimizer.py (hunk 7 above)
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/steppable_optimizer.py:8]: Do I typehint it to just be a callable?

File: qiskit/algorithms/optimizers/steppable_optimizer.py (hunk 8 above)
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/steppable_optimizer.py:22]: Okay, I thought it was only when there were only 2 options. Good to know!

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+        for _ in range(20):
+            ask_object = optimizer.ask()
+            evaluated_gradient = None
+
+            while evaluated_gradient is None:
+                evaluated_gradient = grad(ask_object.x_center)
+                optimizer._state.njev += 1
+
+            tell_object = GD_TellObject(gradient=evaluated_gradient)
```
[the remark this answers is not in the document]
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/steppable_optimizer.py:104]: No, good catch.

File: test/python/algorithms/optimizers/test_gradient_descent.py [hunk 13]
```diff
@@ -59,31 +61,75 @@ def objective(x):
...
 
-        result = optimizer.optimize(circuit.num_parameters, objective, initial_point=initial_point)
+        result = optimizer.minimize(circuit.num_parameters, objective, initial_point=initial_point)
 
         self.assertLess(result[1], -0.95)  # final loss
         self.assertEqual(result[2], 100)  # function evaluations
 
-    def test_callback(self):
-        """Test the callback."""
+    # def test_callback(self):
```
[the remark this answers is not in the document]
Comment (MarcDrudis) [reply test/python/algorithms/optimizers/test_gradient_descent.py:69]: Yes, I just commented it for now because I can't make it work.

File: qiskit/algorithms/optimizers/gradient_descent.py [hunk 14]
```diff
@@ -149,60 +60,104 @@ def settings(self) -> Dict[str, Any]:
...
-                break
+        self._state = GD_OptimizerState(
+            fun=fun,
+            jac=jac,
+            x=np.asarray(x0),
+            nit=0,
+            nfev=0,
+            njev=0,
+            eta=eta,
+            stepsize=np.inf,
```
[the remark this answers is not in the document]
Comment (MarcDrudis) [reply qiskit/algorithms/optimizers/gradient_descent.py:207]: I have changed it now. The reason I did it was to avoid computing an extra bool at every step. I guess this won't be the most expensive part of the evaluation anyways, so it shouldn't matter too much.
